projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/machO/MachOFile.java

```diff
@@ -0,0 +1,289 @@
+        this.header = header;
+        this.loadCommandTable = loadCommandTable;
+        this.buffer = buffer;
+    }
+
+    public List<String> getDyLibs() {
+        List<String> dylibs = new ArrayList<>();
+        for (MachOLoadCommand lc : loadCommandTable.getLoadCommands()) {
+            if (lc instanceof MachODylibCommand) {
+                dylibs.add(((MachODylibCommand) lc).getName());
+            }
+        }
+        return dylibs;
+    }
+
+    public ByteBuffer extractBitcode() {
+
+        ByteBuffer bc = null;
+
+        switch (header.getFileType()) {
+            case MH_OBJECT:
+                bc = extractBitcodeFromObject();
+                break;
+            case MH_EXECUTE:
+                if (loadCommandTable.getSegment(WLLVM_BITCODE_SECTION) != null) {
+                    bc = extractBitcodeFromWLLVMExecute();
+                } else {
+                    bc = extractBitcodeFromExecute();
+                }
+                break;
+            default:
+                throw new RuntimeException("Mach-O file type not supported!");
+        }
+
+        bc.order(ByteOrder.LITTLE_ENDIAN);
+        return bc;
+    }
+
+    private ByteBuffer extractBitcodeFromExecute() {
+        ByteBuffer sectionData = getSectionData(LLVM_SEGMENT, BUNDLE_SECTION);
+        Xar xar = Xar.create(sectionData);
+
+        Document bcDoc = xar.getToc().getTableOfContents();
+        NodeList dataNodes = bcDoc.getElementsByTagName("data");
+
+        // extract all bitcodes from the archive
+        ByteBuffer[] bitcodes = new ByteBuffer[dataNodes.getLength()];
+        for (int i = 0; i < dataNodes.getLength(); i++) {
+            bitcodes[i] = extractBcFromXar(xar, dataNodes.item(i));
+        }
+
+        if (bitcodes.length == 1) {
+            return bitcodes[0];
+        } else {
+            try {
+                return linkBitcodes(bitcodes);
+            } catch (IOException | InterruptedException e) {
+                throw new RuntimeException("Error while attempting to extract bitcode!");
+            }
+        }
+
+    }
+
+    // dummy approach to support regularly built Mach-O files
+    private static ByteBuffer linkBitcodes(ByteBuffer[] bitcodes) throws IOException, InterruptedException {
+
+        // write the bitcodes to the disk
+        List<File> files = new ArrayList<>();
+        for (int i = 0; i < bitcodes.length; i++) {
+            File file = Files.createTempFile(".bc" + i + "_", ".bc").toFile();
+            file.deleteOnExit();
+            FileChannel out = new FileOutputStream(file).getChannel();
+            files.add(file);
+            out.write(bitcodes[i]);
+            out.close();
+        }
+
+        // create llvm-link command
+        File bcFile = Files.createTempFile(".linkedBc_", ".bc").toFile();
+        bcFile.deleteOnExit();
+        StringBuilder cmd = new StringBuilder("llvm-link -o ");
```

Comment:
I don't think its a good idea to call `llvm-link` directly from java. This doesn't check version compatibility like `mx` does which could lead to problems when homebrew updates to a newer llvm version. As a short-term solution we could just throw an exception in this case, telling users to recompile with WLLVM. Also, since we already support libraries, I think it would be possible to just pass each of the `ByteBuffer` to the parser sequentially. We may need to implement logic to find / keep the program entry point in this case, but I think this would be preferable over having this dependency.